BUG: fix choose refcount leak

tylerjereddy opened the pull request:
* Fixes #22683

* use `copyswap` to avoid the reference count leaking reported above when `np.choose` is used with `out`

* my impression from the ticket is that Sebastian doesn't think `copyswap` is a perfect solution, but may suffice short-term?

tylerjereddy committed:
BUG: fix choose refcount leak

* Fixes #22683

* use `copyswap` to avoid the reference count leaking
reported above when `np.choose` is used with `out`

* my impression from the ticket is that Sebastian doesn't
think `copyswap` is a perfect solution, but may suffice
short-term?

tylerjereddy commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -10029,3 +10029,12 @@ def test_argsort_int(N, dtype):
+    np.choose(np.zeros(10000, dtype=int), [a], out=a)
+    np.choose(np.zeros(10000, dtype=int), [a], out=a)
+    refc_end = sys.getrefcount(1)
+    assert refc_end - refc_start < 10
```
could probably go less than `10` here, though I think each incantation can at least do `+1`

seberg commented:
Thanks, we are trying a bit to avoid copyswap; unfortunately, I admit, it would be nice if the replacement was slightly shorter, e.g. see: gh-24176

Happy to do this first, though!   Is this much slower for e.g. `float64`?  (My guess is no, because `memmov` is also just a function call, this loop should be specialized for the typical small itemsizes without references (i.e. basic numeric types).

ngoldbaum commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -1074,12 +1077,11 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
             }
         }
-        memmove(ret_data, PyArray_MultiIter_DATA(multi, mi), elsize);
+        copyswap(ret_data, PyArray_MultiIter_DATA(multi, mi), swap, NULL);
```
I'd prefer not adding new `copyswap` usages if we can avoid it since it will cause seg faults with new dtypes. Happy to push a fixup here if you don't have time to look at the other PRs replacing copyswap with a single-element cast.

tylerjereddy committed:
BUG: PR 24188 revisions

* remove copyswap, carefully controlling
the reference counting to pass the testsuite

tylerjereddy commented:
This PR has been revised based on reviewer feedback to replace `copyswap` with the more verbose approach that tolerates new dtype stuff. It is substantially harder to read, but the CI, including the new regression test, seem to be passing.

ngoldbaum reviewed (approved):
Agreed this is a lot of boilerplate. It probably makes sense to have a dtype API slot that implements a single-element copy, then you wouldn’t need to deal with the flags or the strides (i think?). You’d still need two stages to get the copy function from the dtype and possibly set up the auxdata, but we could certainly simplify this sort of thing a bunch.

Anyway, thanks for rewriting this in a more forward-compatible way while we wait on polishing out the rough edges in the new dtype API.

seberg commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -1074,12 +1074,34 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
+                        &transfer_flags);
+            cast_info.func(&cast_info.context, args, &one,
+                                transfer_strides, cast_info.auxdata);
+        }
```
The boilerplate should really be outside the inner-loop, though (I guess the inputs may be unaligned, in which case we may have to use unaligned if any of them are, I don't think the output can be).  (Not sure that should be dealt with in the core, but fine.)

Also we can branch here based on references or not and remove the `PyArray_INCREF` at the end.
(If we want speed, we need to remove the `memmove` with an itemsize specialized version (like `npy_fasttake_impl`).

(I do think this whole function could probably even use the new iterator without getting too much more unwieldy, considering how unwieldy the setup is, and probably much faster)

charris commented:
@seberg Sounds like this can be improved, but is it good enough for now?

seberg commented:
Just noticed it's also missing the xfree.  I am happy with just moving it out of the inner-loop, but I do think that needs to be done.  I can do that though if @tylerjereddy prefers.

tylerjereddy commented:
I did try hoisting it up last week but got segfaults, will check again today.

tylerjereddy committed:
MAINT: PR 24188 revisions

* hoist the special `out` handling code out of the inner
loop (to the degree the testsuite allowed me to)

* add a missing `NPY_cast_info_xfree`

* adjust the regression test such that it fails before/passes
after on both Python 3.11 and 3.12 beta 4, to deal with
PEP 683

tylerjereddy commented:
Latest set of changes:

    * hoist the special `out` handling code out of the inner
    loop (to the degree the testsuite allowed me to)
    
    * add a missing `NPY_cast_info_xfree`
    
    * adjust the regression test such that it fails before/passes
    after on both Python 3.11 and 3.12 beta 4, to deal with
    PEP 683

seberg committed:
MAINT: Use explicit copy path in choose based on refcheck

seberg committed:
BUG: Remove unnecessary (and now also segfaulting) iterator creation

Also hoist the `dtype` definition up and use it.

seberg committed:
ENH: Use `memcpy` and swap order

Not sure this makes a difference, but we check for memory overlap
so `memmov` isn't necessary and if the compiler keeps the order
intact, we want the `memcpy` path to be the hot one.

seberg commented:
Thanks I added three commits to:
1. Deal with errors during the copy (which are not possible for NumPy dtypes copies, but in principle are possible, e.g. for Nathan's strings).
2. Use this always if there are references:
   * Avoids making `out=` special, which IMO shouldn't be different really
   * New dtypes don't support `PyArray_INCREF()` maybe they never will since the API really isn't particularly clean (and may be outright impossible for some dtypes or `object` arrays in alternative Python implementations).
3. Remove stray iterator creation (and fix to use `obj` rather than `out` for alignment, which went from irrelevant bug to a crash due to my changes).
4. Use `memcpy` rather than `memmov`. We check for no memory overlap, there really isn't a reason to not use it, so I think it is clearer for future iterations also (and probably faster).

Also thought it would be clearer to use `dtype` since it is defined anyway.

ngoldbaum commented:
Looks good to me, going to merge. Thanks for pushing this through @tylerjereddy and getting it across the line @seberg!
